### PR TITLE
Feature/BNGP-4840-add-planning-authority-and-references-to-allocation-journey:

### PR DESCRIPTION
**Pull request by Stewart-Jumbe**
Allocation: Add Planning Authority and References
https://eaflood.atlassian.net/browse/BNGP-4840
add new page 'Development project details', allows users to add development project name, local planning authority and planning application reference

### CODE REVIEW

File: packages/webapp/src/utils/developer-constants.js [hunk 1]
```diff
@@ -88,6 +90,9 @@ const DEVELOPER_PLANNING_DECISION_FILE_SIZE = 'planning-decision-file-siz'
 const DEVELOPER_PLANNING_DECISION_FILE_TYPE = 'planning-decision-file-type'
 const DEVELOPER_PLANNING_DECISION_FILE_NAME = 'planning-decision-file-name'
 const DEVELOPER_PLANNING_DECISION_CHECKED = 'planning-decision-checked'
+const DEVELOPER_PLANNING_AUTHORITY_LIST = 'planning-authority-list'
+const DEVELOPER_PLANNING_APPLICATION_REF = 'planning-application-ref'
+const DEVELOPER_DEVELOPMENT_NAME = 'planning-development-name'
```
Comment (paulsimonandrews) [inline packages/webapp/src/utils/developer-constants.js:95]: These three keys should have `developer-` prefixed to them (e.g. `developer-planning-development-name`) to avoid clashes with registration journey keys

File: packages/webapp/src/utils/developer-constants.js [hunk 2]
```diff
@@ -1,44 +1,46 @@
...
-const DEVELOPER_CHECK_WRITTEN_AUTHORISATION_FILE = 'developer/check-written-authorisation-file'
-const DEVELOPER_CHECK_PLANNING_DECISION_NOTICE_FILE = 'developer/check-planning-decision-notice-file'
-const DEVELOPER_UPLOAD_PLANNING_DECISION_NOTICE = 'developer/upload-planning-decision-notice'
-const DEVELOPER_DOWNLOAD_WRITTEN_AUTHORISATION = 'developer/download-written-authorisation-file'
-const DEVELOPER_DOWNLOAD_PLANNING_DECISION_FILE = 'developer/download-planning-decision-file'
-const DEVELOPER_DEFRA_ACCOUNT_NOT_LINKED = 'developer/defra-account-not-linked'
-const DEVELOPER_UPLOAD_CONSENT_TO_ALLOCATE_GAINS = 'developer/upload-consent-to-allocate-gains'
-const DEVELOPER_CHECK_CONSENT_TO_USE_GAIN_SITE_FILE = 'developer/check-consent-file'
-const DEVELOPER_PLANNING_DECISION_UPLOAD = 'developer/planning-decision-upload'
+const routes = {
```
Comment (paulsimonandrews) [inline packages/webapp/src/utils/developer-constants.js:2]: Revert to the old way of doing route constants

Allocation and Registration differ in the way constants are used than Credits Purchase (for historical reasons), so we should keep Allocation consistent and follow its conventions for this PR

File: packages/webapp/src/utils/developer-constants.js [hunk 3]
```diff
@@ -1,44 +1,46 @@
...
+  DEVELOPER_DOWNLOAD_PLANNING_DECISION_FILE: 'developer/download-planning-decision-file',
+  DEVELOPER_DEFRA_ACCOUNT_NOT_LINKED: 'developer/defra-account-not-linked',
+  DEVELOPER_UPLOAD_CONSENT_TO_ALLOCATE_GAINS: 'developer/upload-consent-to-allocate-gains',
+  DEVELOPER_CHECK_CONSENT_TO_USE_GAIN_SITE_FILE: 'developer/check-consent-file',
+  DEVELOPER_PLANNING_DECISION_UPLOAD: 'developer/planning-decision-upload',
+  DEVELOPER_DEVELOPMENT_PROJECT_INFORMATION: '/developer/development-project-information'
+}
 
 // ./Routes constants
+const DEVELOPER_IS_AGENT = 'developer-is-agent'
```
Comment (paulsimonandrews) [inline packages/webapp/src/utils/developer-constants.js:43]: Put with the other RedisKey constants

File: packages/webapp/src/utils/developer-constants.js [hunk 4]
```diff
@@ -127,46 +132,12 @@ const ROUTING_REGISTER_OPTIONS = {
   RECORD: 'record-off-site'
 }
 
+const views = Object.fromEntries(
```
Comment (paulsimonandrews) [inline packages/webapp/src/utils/developer-constants.js:135]: This isn't needed as the final route and views are created and exported in `constants.js` and we don't use `developer-constants.js` directly

File: packages/webapp/src/utils/developer-constants.js [hunk 5]
```diff
@@ -194,11 +165,11 @@ export default {
...
     DEVELOPER_DEFRA_ACCOUNT_DETAILS_CONFIRMED,
-    DEVELOPER_LANDOWNER_OR_LEASEHOLDER,
+    DEVELOPER_LANDOWNER_OR_LEASEHOLDER: routes.DEVELOPER_LANDOWNER_OR_LEASEHOLDER,
     DEVELOPER_LANDOWNER_TYPE,
-    DEVELOPER_CLIENT_INDIVIDUAL_ORGANISATION,
-    DEVELOPER_CLIENTS_NAME,
-    DEVELOPER_CLIENTS_ORGANISATION_NAME,
+    DEVELOPER_CLIENT_INDIVIDUAL_ORGANISATION: routes.DEVELOPER_CLIENT_INDIVIDUAL_ORGANISATION,
+    DEVELOPER_CLIENTS_NAME: routes.DEVELOPER_CLIENTS_NAME,
+    DEVELOPER_CLIENTS_ORGANISATION_NAME: routes.DEVELOPER_CLIENTS_ORGANISATION_NAME,
```
Comment (paulsimonandrews) [inline packages/webapp/src/utils/developer-constants.js:172]: If you revert the changes to the routes above, won't need these redis keys defined this way

File: packages/webapp/src/utils/developer-task-list.js [hunk 6]
```diff
@@ -44,6 +44,21 @@ const developerTaskList = {
         }
       ]
     },
+    {
+      taskTitle: 'Development project information',
```
Comment (paulsimonandrews) [inline packages/webapp/src/utils/developer-task-list.js:48]: Changes to this file won't be needed. This file has been removed from the master branch now.

File: packages/webapp/src/journey-validation/allocation/development-project-information.js [hunk 7]
```diff
@@ -0,0 +1,22 @@
+import developerConstants from '../../utils/developer-constants.js'
```
Comment (paulsimonandrews) [inline packages/webapp/src/journey-validation/allocation/development-project-information.js:1]: Use `constants.js` rather than `developer-constants.js`

File: packages/webapp/src/journey-validation/allocation/task-sections.js [hunk 8]
```diff
@@ -1,10 +1,11 @@
 import constants from '../../utils/constants.js'
 import { taskDefinition, taskSectionDefinition } from '../utils.js'
 import { applicantDetailsJourneys } from './applicant-details.js'
-import { confirmDevelopmentHabitatDetailsJourneys } from './confirm-development-habitat-details.js'
 import { bngNumberJourneys } from './biodiversity-net-gain-number.js'
 import { addMetricCalculationsJourneys } from './add-metric-calculations.js'
 import { planningDecisionNoticeJourneys } from './planning-decision-notice.js'
+import { addDevelopmentProjectInformationJourneys } from './development-project-information.js'
+import developerConstants from '../../utils/developer-constants.js'
```
Comment (paulsimonandrews) [inline packages/webapp/src/journey-validation/allocation/task-sections.js:8]: Use `constants.js` rather than `developer-constants.js`

File: packages/webapp/src/journey-validation/allocation/task-sections.js [hunk 9]
```diff
@@ -14,6 +15,14 @@ const applicantDetails = taskDefinition(
   applicantDetailsJourneys
 )
 
+const addDevlopmentProjectInformation = taskDefinition(
+  'add-devlopment-project-information',
+  'Add development project information',
+  developerConstants.routes.DEVELOPER_DEVELOPMENT_PROJECT_INFORMATION,
+  developerConstants.routes.DEVELOPER_DEVELOPMENT_PROJECT_INFORMATION,
```
Comment (paulsimonandrews) [inline packages/webapp/src/journey-validation/allocation/task-sections.js:22]: Use `constants.` rather than `developerConstants`

File: packages/webapp/src/routes/developer/development-project-information.js [hunk 10]
```diff
@@ -0,0 +1,113 @@
+import developerConstants from '../../utils/developer-constants.js'
```
Comment (paulsimonandrews) [inline packages/webapp/src/routes/developer/development-project-information.js:1]: Use `constants.js` rather than `developer-constants.js`

And change below to use `constants.` rather than `developerConstants.`

File: packages/webapp/src/routes/developer/development-project-information.js [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+const handlers = {
+  get: (request, h) => {
+    const lpaNames = getLpaNames(filePathAndName)
+
+    request.yar.set(constants.redisKeys.REF_LPA_NAMES, lpaNames)
+
+    const selectedLpa = request.yar.get(developerConstants.redisKeys.DEVELOPER_PLANNING_AUTHORITY_LIST)
+    const planningApplicationRef = request.yar.get(developerConstants.redisKeys.DEVELOPER_PLANNING_APPLICATION_REF)
+    const developmentName = request.yar.get(developerConstants.redisKeys.DEVELOPER_DEVELOPMENT_NAME)
+    console.log(developerConstants)
```
Comment (paulsimonandrews) [inline packages/webapp/src/routes/developer/development-project-information.js:19]: Remove console log

File: packages/webapp/src/utils/developer-constants.js [hunk 12]
```diff
@@ -1,44 +1,46 @@
...
+  DEVELOPER_CHECK_WRITTEN_AUTHORISATION_FILE: 'developer/check-written-authorisation-file',
+  DEVELOPER_CHECK_PLANNING_DECISION_NOTICE_FILE: 'developer/check-planning-decision-notice-file',
+  DEVELOPER_UPLOAD_PLANNING_DECISION_NOTICE: 'developer/upload-planning-decision-notice',
+  DEVELOPER_DOWNLOAD_WRITTEN_AUTHORISATION: 'developer/download-written-authorisation-file',
+  DEVELOPER_DOWNLOAD_PLANNING_DECISION_FILE: 'developer/download-planning-decision-file',
+  DEVELOPER_DEFRA_ACCOUNT_NOT_LINKED: 'developer/defra-account-not-linked',
+  DEVELOPER_UPLOAD_CONSENT_TO_ALLOCATE_GAINS: 'developer/upload-consent-to-allocate-gains',
+  DEVELOPER_CHECK_CONSENT_TO_USE_GAIN_SITE_FILE: 'developer/check-consent-file',
+  DEVELOPER_PLANNING_DECISION_UPLOAD: 'developer/planning-decision-upload',
+  DEVELOPER_DEVELOPMENT_PROJECT_INFORMATION: '/developer/development-project-information'
```
Comment (paulsimonandrews) [inline packages/webapp/src/utils/developer-constants.js:39]: Remove `/` from beginning of route

**Review by paulsimonandrews — CHANGES_REQUESTED** [review]
Specific comments in the code.

Move to using `constants.js` rather than the developer constants file to follow the convention used on allocation journey.

Behaviour of the page and update to task list look good, so just need to get the structure of the code in line.

Also please resolve conflicts and get the branch up to date with master. Hopefully this will trigger the pipeline to run the tests and Sonar analysis.

File: packages/webapp/src/utils/developer-constants.js (hunk 12 above)
Comment (Stewart-Jumbe) [reply packages/webapp/src/utils/developer-constants.js]: done

File: packages/webapp/src/routes/developer/development-project-information.js (hunk 11 above)
Comment (Stewart-Jumbe) [reply packages/webapp/src/routes/developer/development-project-information.js]: Done

File: packages/webapp/src/routes/developer/development-project-information.js (hunk 10 above)
Comment (Stewart-Jumbe) [reply packages/webapp/src/routes/developer/development-project-information.js]: Done

File: packages/webapp/src/journey-validation/allocation/task-sections.js (hunk 9 above)
Comment (Stewart-Jumbe) [reply packages/webapp/src/journey-validation/allocation/task-sections.js]: Done

File: packages/webapp/src/journey-validation/allocation/task-sections.js (hunk 8 above)
Comment (Stewart-Jumbe) [reply packages/webapp/src/journey-validation/allocation/task-sections.js]: Done

File: packages/webapp/src/utils/developer-constants.js (hunk 5 above)
Comment (Stewart-Jumbe) [reply packages/webapp/src/utils/developer-constants.js]: Done

File: packages/webapp/src/utils/developer-constants.js (hunk 4 above)
Comment (Stewart-Jumbe) [reply packages/webapp/src/utils/developer-constants.js]: done

File: packages/webapp/src/utils/developer-constants.js (hunk 1 above)
Comment (Stewart-Jumbe) [reply packages/webapp/src/utils/developer-constants.js]: done

File: packages/webapp/src/utils/developer-constants.js (hunk 2 above)
Comment (Stewart-Jumbe) [reply packages/webapp/src/utils/developer-constants.js]: done

File: packages/webapp/src/utils/developer-constants.js (hunk 3 above)
Comment (Stewart-Jumbe) [reply packages/webapp/src/utils/developer-constants.js]: done

File: packages/webapp/src/journey-validation/allocation/task-sections.js [hunk 13]
```diff
@@ -58,11 +58,11 @@ const checkYourAnswers = {
 
 const taskSections = [
   taskSectionDefinition('Applicant information', [applicantDetails]),
+  taskSectionDefinition('Development project information', [addDevlopmentProjectInformation]),
```
Comment (paulsimonandrews) [inline packages/webapp/src/journey-validation/allocation/task-sections.js:61]: This section should be in the 'Development information' section below. So remove this line and add `addDevlopmentProjectInformation` to the below section

File: packages/webapp/src/journey-validation/allocation/task-sections.js [hunk 14]
```diff
@@ -58,11 +58,11 @@ const checkYourAnswers = {
...
   taskSectionDefinition('Development information', [
     planningDecisionNotice,
     biodiversityGainSiteNumber,
-    biodiversityMetricCalculations,
-    confirmDevelopmentHabitatDetails
+    biodiversityMetricCalculations
```
Comment (paulsimonandrews) [inline packages/webapp/src/journey-validation/allocation/task-sections.js:65]: Add `addDevlopmentProjectInformation` into here

**Review by paulsimonandrews — CHANGES_REQUESTED** [review]
All looks good, just need to move the section in the task list

**Comment by shaundefra** [comment]
tested locallly, bug to do with error messaging raised separately
